### ISSUE TITLE
Problem with OpenRC cgroup inside a container since LXC 2.x

### ISSUE THREAD

**Issue by jirutka** [lxc/lxc#1095]
I’ve tried to upgrade from lxc 1.1.5 to 2.0.3 on Alpine Linux and ran into some problem with cgroups. It seems that cgroup mount `/sys/fs/cgroup/openrc` is now somehow propagated from the host into a container and causes a conflict with OpenRC running inside the container.
### lxc 2.0.3

```
host:~# lxc-start -n test -F

   OpenRC 0.21.723ca9e is starting up Linux 4.4.15 (x86_64) [LXC]

 * /proc is already mounted
 * Mounting /run ... * /run/openrc: creating directory
 * /run/lock: creating directory
 * /run/lock: correcting owner
 * Caching service dependencies ... [ ok ]
mkdir: can't create directory '/sys/fs/cgroup/openrc/bootmisc': Read-only file system
mkdir: can't create directory '/sys/fs/cgroup/openrc/bootmisc': Read-only file system
 * Creating user login records ... [ ok ]
 * Wiping /tmp directory ... [ ok ]
mkdir: can't create directory '/sys/fs/cgroup/openrc/hosts': Read-only file system
mkdir: can't create directory '/sys/fs/cgroup/openrc/hosts': Read-only file system
mkdir: can't create directory '/sys/fs/cgroup/openrc/syslog': Read-only file system
mkdir: can't create directory '/sys/fs/cgroup/openrc/syslog': Read-only file system
 * Starting busybox syslog ... [ ok ]
mkdir: can't create directory '/sys/fs/cgroup/openrc/networking': Read-only file system
mkdir: can't create directory '/sys/fs/cgroup/openrc/networking': Read-only file system
...
```

Let’s look what we have there (inside the container):

```
test: ~# mount | grep /sys/fs/cgroup
cgroup_root on /sys/fs/cgroup type tmpfs (rw,nosuid,nodev,noexec,relatime,size=10240k,mode=755)
cgroup_root on /sys/fs/cgroup/openrc type tmpfs (ro,relatime,size=10240k,mode=755)
openrc on /sys/fs/cgroup/openrc/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,release_agent=/lib/rc/sh/cgroup-release-agent.sh,name=openrc)
cgroup_root on /sys/fs/cgroup/cpuset type tmpfs (ro,relatime,size=10240k,mode=755)
cpuset on /sys/fs/cgroup/cpuset/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset,clone_children)
cgroup_root on /sys/fs/cgroup/cpu type tmpfs (ro,relatime,size=10240k,mode=755)
cpu on /sys/fs/cgroup/cpu/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup_root on /sys/fs/cgroup/cpuacct type tmpfs (ro,relatime,size=10240k,mode=755)
cpuacct on /sys/fs/cgroup/cpuacct/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpuacct)
cgroup_root on /sys/fs/cgroup/blkio type tmpfs (ro,relatime,size=10240k,mode=755)
blkio on /sys/fs/cgroup/blkio/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup_root on /sys/fs/cgroup/memory type tmpfs (ro,relatime,size=10240k,mode=755)
memory on /sys/fs/cgroup/memory/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup_root on /sys/fs/cgroup/devices type tmpfs (ro,relatime,size=10240k,mode=755)
devices on /sys/fs/cgroup/devices/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup_root on /sys/fs/cgroup/freezer type tmpfs (ro,relatime,size=10240k,mode=755)
freezer on /sys/fs/cgroup/freezer/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup_root on /sys/fs/cgroup/net_cls type tmpfs (ro,relatime,size=10240k,mode=755)
net_cls on /sys/fs/cgroup/net_cls/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls)
cgroup_root on /sys/fs/cgroup/net_prio type tmpfs (ro,relatime,size=10240k,mode=755)
net_prio on /sys/fs/cgroup/net_prio/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,net_prio)
cgroup_root on /sys/fs/cgroup/pids type tmpfs (ro,relatime,size=10240k,mode=755)
pids on /sys/fs/cgroup/pids/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,pids)

test:~# tree /sys/fs/cgroup/openrc/
/sys/fs/cgroup/openrc/
└── lxc
    └── test
        ├── cgroup.clone_children
        ├── cgroup.procs
        ├── notify_on_release
        └── tasks
```

Directory `/sys/fs/cgroup/openrc` is indeed not writable, because there’s already `cgroup_root` mounted as read-only.
### lxc 1.1.5

Exactly the same rootfs and configuration on lxc 1.1.5:

```
host:~# lxc-start -n test -F

   OpenRC 0.21.723ca9e is starting up Linux 4.4.15 (x86_64) [LXC]

 * /proc is already mounted
 * Mounting /run ... * /run/openrc: creating directory
 * /run/lock: creating directory
 * /run/lock: correcting owner
 * Caching service dependencies ... [ ok ]
 * Creating user login records ... [ ok ]
 * Wiping /tmp directory ... [ ok ]
 * Starting busybox syslog ... [ ok ]
...
```

```
test:~# mount | grep /sys/fs/cgroup
cgroup_root on /sys/fs/cgroup type tmpfs (rw,nosuid,nodev,noexec,relatime,size=10240k,mode=755)
cgroup_root on /sys/fs/cgroup/pids type tmpfs (ro,relatime,size=10240k,mode=755)
pids on /sys/fs/cgroup/pids/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,pids)
cgroup_root on /sys/fs/cgroup/net_prio type tmpfs (ro,relatime,size=10240k,mode=755)
net_prio on /sys/fs/cgroup/net_prio/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,net_prio)
cgroup_root on /sys/fs/cgroup/net_cls type tmpfs (ro,relatime,size=10240k,mode=755)
net_cls on /sys/fs/cgroup/net_cls/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,net_cls)
cgroup_root on /sys/fs/cgroup/freezer type tmpfs (ro,relatime,size=10240k,mode=755)
freezer on /sys/fs/cgroup/freezer/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
cgroup_root on /sys/fs/cgroup/devices type tmpfs (ro,relatime,size=10240k,mode=755)
devices on /sys/fs/cgroup/devices/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup_root on /sys/fs/cgroup/memory type tmpfs (ro,relatime,size=10240k,mode=755)
memory on /sys/fs/cgroup/memory/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup_root on /sys/fs/cgroup/blkio type tmpfs (ro,relatime,size=10240k,mode=755)
blkio on /sys/fs/cgroup/blkio/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup_root on /sys/fs/cgroup/cpuacct type tmpfs (ro,relatime,size=10240k,mode=755)
cpuacct on /sys/fs/cgroup/cpuacct/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpuacct)
cgroup_root on /sys/fs/cgroup/cpu type tmpfs (ro,relatime,size=10240k,mode=755)
cpu on /sys/fs/cgroup/cpu/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup_root on /sys/fs/cgroup/cpuset type tmpfs (ro,relatime,size=10240k,mode=755)
cpuset on /sys/fs/cgroup/cpuset/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset,clone_children)

test:~# ls /sys/fs/cgroup/
blkio     cpu       cpuacct   cpuset    devices   freezer   memory    net_cls   net_prio  pids
```

As you can see, `/sys/fs/cgroup/openrc` is not mounted here.

``` diff
-cgroup_root on /sys/fs/cgroup/openrc type tmpfs (ro,relatime,size=10240k,mode=755)
-openrc on /sys/fs/cgroup/openrc/lxc/test type cgroup (rw,nosuid,nodev,noexec,relatime,release_agent=/lib/rc/sh/cgroup-release-agent.sh,name=openrc)
```
### Notes
- The container’s rootfs is created using the download template in lxc 2.0.3, i.e. from prebuilt LXC image, with default configuration.
- Host system, container’s rootfs and lxc configuration is exactly the same (I’ve just removed `lxc.rootfs.backend = dir` that is not supported in lxc 1.1.5), the only difference between the above examples is the lxc version.
- To be sure that it’s not related to grsecurity, I’m running vanilla kernel.
- I have set `memory.use_hierarchy` to 1.
- I do _not_ have cgmanager installed.

/cc @ncopa

**Comment by brauner** [comment]
This seems like `cgfsng` assumes that `/sys/fs/cgroup/openrc` is a valid cgroup controller when it actually isn't. Are cgroups on `Alpine` mounted under `/sys/fs/cgroup/*`? And `/sys/fs/cgroup/openrc` is a separate mount specific to `Alpine` which has nothing to do with cgroups?

**Comment by brauner** [comment]
Oh, nevermind, openrc is a valid cgroup of your init system.

**Comment by jirutka** [comment]
Yes, it's a valid cgroup of OpenRC, and yes, Alpine mounts cgroups under `/sys/fs/cgroup/` (see above outputs of `mount` command). What I can do with it?

The same problem will probably affect even Gentoo and other distros using OpenRC.

**Comment by jirutka** [comment]
I’ve tried to set `lxc.cgroup.use` and… it behaves really weird.

`lxc.cgroup.use = blkio,cpu,cpuacct,cpuset,devices,freezer,memory,net_cls,net_prio,pids`
`lxc.cgroup.use =`
`lxc.cgroup.use = @all`

→ No difference, behaves the same as without this setting.

`lxc.cgroup.use = foo,blkio,cpu,cpuacct,cpuset,devices,freezer,memory,net_cls,net_prio,pids` (note invalid cgroup `foo`)

→ It magically **works**, all cgroups that should be mounted are really mounted and `/sys/fs/cgroup/openrc` is not, i.e. no errors from OpenRC. How is that possible?! o.O

`lxc.cgroup.use = foo`

→ Does not work (obviously), fails with:

```
lxc-start: start.c: lxc_spawn: 1100 failed creating cgroups
lxc-start: start.c: __lxc_start: 1353 failed to spawn 'test'
lxc-start: lxc_start.c: main: 344 The container failed to start.
lxc-start: lxc_start.c: main: 348 Additional information can be obtained by setting the --logfile and --logpriority options.
```

**Comment by jirutka** [comment]
I found in the sources ([cgfs.c:583](https://github.com/lxc/lxc/blob/5429ec248a1090edde6285e4bc729641f6f79b57/src/lxc/cgfs.c#L583)) that `lxc.cgroup.use` also accepts special value `@kernel`. Why is this not documented anywhere?! Also `@all` is not in the documentation, I found it when googling `lxc.cgroup.use`.

Anyway, adding `lxc.cgroup.use = @kernel` to `/etc/lxc/lxc.conf` fixed the problem! Now it works in the same way as in 1.1.5, i.e. all cgroups are mounted, except `openrc`.

**Comment by brauner** [comment]
I'm looking into this. `@all`, `@named`, `@kernel` seem to be implemented only in `cgfs.c` so far and not in `cgfsng.c`. Can you start your container with `-l DEBUG -o alpine_log` and append the `alpine_log` file here?

**Comment by brauner** [comment]
This is as I expected. With `lxc.cgroup.use` you use `cgfsng.c` but with `lxc.cgroup.use = @kernel` you're using `cgfs.c`. (More a note for myself.)

**Comment by brauner** [comment]
I reckon that `lxc.cgroup.use` is not fully implemented in `cgfsng.c`. /cc @hallyn
Have the two container, of which you appended the logs, been created and started identically aka do they only differ with respect of `lxc.cgroup.use`. Also, please provide the version of all lxc packages installed. :)

**Comment by jirutka** [comment]
What do you mean with “all lxc packages”? LXC is just one package, isn’t it? I already wrote it, lxc 2.0.3 installed from [this](https://pkgs.alpinelinux.org/package/edge/main/x86_64/lxc) Alpine package.

**Comment by brauner** [comment]
Ok, most distros ship `lxc` in separate packages e.g. `lxc`, `liblxc`, `liblxc-devel` a version mismatch could have been informative.

**Comment by jirutka** [comment]
Alpine also ships lxc in multiple packages (lxc, lxc-templates, lxc-libs, lua5.2-lxc, lxc-doc, lxc-dev, lxc-lvm), but all of them are (obviously) built from the same upstream version and I have also installed correct versions. :)

**Comment by brauner** [comment]
- About the `mkdir` errors.
  - `cgfsng.c` and `cgfs.c` behave differently: `cgfs.c` will not mount named subsystems per default. It will only mount named subsystems when  `lxc.cgroup.use = @all`. If I set this options and `cgfs.c` is used (If this is the case can be verified in the logs.) I can reproduce the error you only seem to see with `cgfsng.c`. The reason for this is that `/sys/fs/cgroup/openrc` is mounted `ro` because `lxc.mount.auto = cgroup:mixed` is set. If you want to prevent this error, then you can set `lxc.mount.auto = cgroup:rw` in the containers config. This will mount the cgroups as `rw` and allow your init to create directories under `/sys/fs/cgroup/openrc`.

- About the strange behavior of `lxc.cgroup.use`
  - Currently `cgfsng.c` does not fully implement this option and we need to think about how we want to do this.

In summary, I'd suggest to close this issue and open a new one about `lxc.cgroup.use` usage with the `cgfsng.c` cgroup driver.

**Comment by jirutka** [comment]
> The reason for this is that `/sys/fs/cgroup/openrc` is mounted ro because `lxc.mount.auto = cgroup:mixed` is set. If you want to prevent this error, then you can set `lxc.mount.auto = cgroup:rw` in the containers config.

I’m really not sure whether `/sys/fs/cgroup/openrc` should be mounted by LXC inside a container at all. Look at the snippet in the first comment, when LXC mounts `/sys/fs/cgroup/openrc`, then it contains:

```
test:~# tree /sys/fs/cgroup/openrc/
/sys/fs/cgroup/openrc/
└── lxc
    └── test
        ├── cgroup.clone_children
        ├── cgroup.procs
        ├── notify_on_release
        └── tasks
```

This is inherited (?) from the host and IMO doesn’t make sense for OpenRC inside the container.

**Comment by brauner** [comment]
> I’m really not sure whether /sys/fs/cgroup/openrc should be mounted by LXC inside a container at all. 

`openrc` is a valid (named) cgroup on your system and we shouldn't special case what cgroups each init system does or does not want to mount. We might think about adding some flexibility about specifying what cgroups to mount per container (*If* that is feasible.). I'm currently not sure about the original motivation for `cgfs.c` to *not* mount named cgroups.

> This is inherited (?) from the host and IMO doesn’t make sense for OpenRC inside the container.

Yes, this is inherited from the host.

**Comment by mcassaniti** [comment]
Has this been tried without the `lxc.cgroup.use = @kernel` option for LXC version 3? Is it reasonable to close?

**Comment by stgraber** [comment]
Pretty sure we can close. Nowadays we have proper namespacing for cgroups and certainly are running openrc in containers just fine.
